framework/src/modules/poa/module.ts

```diff
@@ -31,11 +42,183 @@ export class PoAModule extends BaseModule {
+		if (threshold < totalWeight / BigInt(3) + BigInt(1) || threshold > totalWeight) {
+			throw new Error('`threshold` in snapshot substore is not within range.');
+		}
+
+		// Create an entry in the validator substore for each entry validator in the validators
+		// Create an entry in the name substore for each entry validator in the validators
+		const validatorStore = this.stores.get(ValidatorStore);
+		const nameStore = this.stores.get(NameStore);
+
+		for (const currentValidator of validators) {
+			await validatorStore.set(context, currentValidator.address, { name: currentValidator.name });
+			await nameStore.set(context, Buffer.from(currentValidator.name, 'utf-8'), {
+				address: currentValidator.address,
+			});
+		}
+
+		// Create three entries in the snapshot substore indicating a snapshot of the next rounds of validators
+		const snapshotStore = this.stores.get(SnapshotStore);
+		await snapshotStore.set(context, utils.intToBuffer(0, 4), {
+			...snapshotSubstore,
+			validators: activeValidators,
+		});
+		await snapshotStore.set(context, utils.intToBuffer(1, 4), {
+			...snapshotSubstore,
+			validators: activeValidators,
+		});
+		await snapshotStore.set(context, utils.intToBuffer(2, 4), {
+			...snapshotSubstore,
+			validators: activeValidators,
+		});
+
+		// Create an entry in the chain properties substore
+		const { header } = context;
+		const chainPropertiesStore = this.stores.get(ChainPropertiesStore);
+		await chainPropertiesStore.set(context, EMPTY_BYTES, {
+			roundEndHeight: header.height,
+			validatorsUpdateNonce: 0,
+		});
+	}
+
+	public async finalizeGenesisState(context: GenesisBlockExecuteContext): Promise<void> {
+		const genesisBlockAssetBytes = context.assets.getAsset(MODULE_NAME_POA);
+		if (!genesisBlockAssetBytes) {
+			return;
+		}
+		const asset = codec.decode<GenesisPoAStore>(genesisPoAStoreSchema, genesisBlockAssetBytes);
+		const snapshotStore = this.stores.get(SnapshotStore);
+		const currentRoundSnapshot = await snapshotStore.get(context, utils.intToBuffer(0, 4));
+		const chainPropertiesStore = this.stores.get(ChainPropertiesStore);
+		const chainProperties = await chainPropertiesStore.get(context, EMPTY_BYTES);
+
+		await chainPropertiesStore.set(context, EMPTY_BYTES, {
+			...chainProperties,
+			roundEndHeight: chainProperties.roundEndHeight + currentRoundSnapshot.validators.length,
+		});
+
+		// Pass the required information to the Validators module.
+		const methodContext = context.getMethodContext();
+		await this._validatorsMethod.setValidatorsParams(
+			methodContext,
+			context,
+			currentRoundSnapshot.threshold,
+			currentRoundSnapshot.threshold,
+			currentRoundSnapshot.validators.map(v => ({ address: v.address, bftWeight: v.weight })),
+		);
+
+		// Pass the BLS keys and generator keys to the Validators module.
+		for (const v of asset.validators) {
+			const isValid = await this._validatorsMethod.registerValidatorKeys(
```

Comment:
I've noticed that as well. To me the returning of false when invalid key and true when success seems fine. If not perhaps it can be addressed in another issue. @shuse2

Reply:
Could you please create an issue for SDK to have a look at it? I don't think that this is a serious problem, but this is a protocol difference between the LIP and the implementation, it's better to discuss it.